src/core/labeling/qgspallabeling.cpp

```diff
@@ -2373,119 +2373,159 @@ std::unique_ptr<QgsLabelFeature> QgsPalLayerSettings::registerFeatureWithDetails
     }
   }
 
-  if ( mDataDefinedProperties.isActive( QgsPalLayerSettings::PositionX ) )
+  bool hasDataDefinedPosition = false;
   {
-    exprVal = mDataDefinedProperties.value( QgsPalLayerSettings::PositionX, context.expressionContext() );
-    if ( !exprVal.isNull() )
+    bool ddPosition = false;
+
+    if ( mDataDefinedProperties.isActive( QgsPalLayerSettings::PositionX )
+         && mDataDefinedProperties.isActive( QgsPalLayerSettings::PositionY )
+         && !mDataDefinedProperties.value( QgsPalLayerSettings::PositionX, context.expressionContext() ).isNull()
+         && !mDataDefinedProperties.value( QgsPalLayerSettings::PositionY, context.expressionContext() ).isNull() )
+    {
+      ddPosition = true;
+
+      bool ddXPos = false, ddYPos = false;
+      xPos = mDataDefinedProperties.value( QgsPalLayerSettings::PositionX, context.expressionContext() ).toDouble( &ddXPos );
+      yPos = mDataDefinedProperties.value( QgsPalLayerSettings::PositionY, context.expressionContext() ).toDouble( &ddYPos );
+      if ( ddXPos && ddYPos )
+        hasDataDefinedPosition = true;
+    }
+    else if ( mDataDefinedProperties.isActive( QgsPalLayerSettings::PositionPoint )
+              && !mDataDefinedProperties.value( QgsPalLayerSettings::PositionPoint, context.expressionContext() ).isNull() )
     {
-      if ( !exprVal.isNull() )
-        xPos = exprVal.toDouble( &ddXPos );
+      ddPosition = true;
 
-      if ( mDataDefinedProperties.isActive( QgsPalLayerSettings::PositionY ) )
+      QVariant pointAsVariant = mDataDefinedProperties.value( QgsPalLayerSettings::PositionPoint, context.expressionContext() );
+      QgsPoint point;
+      if ( pointAsVariant.canConvert<QgsReferencedGeometry>() )
       {
-        exprVal = mDataDefinedProperties.value( QgsPalLayerSettings::PositionY, context.expressionContext() );
-        if ( !exprVal.isNull() )
+        point = QgsPoint( pointAsVariant.value<QgsReferencedGeometry>().asPoint() );
+      }
+      else if ( pointAsVariant.canConvert<QgsGeometry>() )
+      {
+        point = QgsPoint( pointAsVariant.value<QgsGeometry>().asPoint() );
+      }
+      else if ( !pointAsVariant.toString().isEmpty() )
+      {
+        point.fromWkt( pointAsVariant.toString() );
+      }
```

Comment:
I don't think this logic is correct. I think if a user has a field with WKT, they should use an expression with the geom_from_wkt() function instead (which would mean that the result is always a QgsGeometry)